api/currency_conversion_test.go

```diff
@@ -0,0 +1,341 @@
+			},
+		})
+	})
+
+	It("does a conversion to GBP for resources billed in USD", func() {
+		guid := "00000001-0000-0000-0000-000000000000"
+		spaceGUID := "00000001-0001-0000-0000-000000000000"
+
+		createScenario(
+			guid,
+			spaceGUID,
+			start,
+			[][]string{
+				{"USD", "$time_in_seconds"},
+			},
+		)
+
+		addCurrency("USD", 0.5, start)
+
+		var out interface{}
+
+		request(
+			"/organisations/"+guid+"/spaces",
+			start,
+			start.Add(2*24*time.Hour),
+			&out,
+		)
+
+		expectedPrice :=
+			2 * // days
+				24 * 3600 * // seconds in a day
+				0.5 * // USD => GBP
+				100 // pences
+
+		ExpectJSON(out, []map[string]interface{}{
+			{
+				"org_guid":               guid,
+				"space_guid":             spaceGUID,
+				"price_in_pence_ex_vat":  expectedPrice,
+				"price_in_pence_inc_vat": expectedPrice * 1.20,
+			},
+		})
+	})
+
+	It("copes with multiple currency entries with same rate", func() {
+		guid := "00000001-0000-0000-0000-000000000000"
+		spaceGUID := "00000001-0001-0000-0000-000000000000"
+
+		createScenario(
+			guid,
+			spaceGUID,
+			start,
+			[][]string{
+				{"USD", "$time_in_seconds"},
+			},
+		)
+
+		addCurrency("USD", 1, start)
+		addCurrency("USD", 1, start.Add(24*1*time.Hour))
+
+		var out interface{}
+
+		request(
+			"/organisations/"+guid+"/spaces",
+			start,
+			start.Add(2*24*time.Hour),
+			&out,
+		)
+
+		expectedPrice :=
+			2 * // days
+				24 * 3600 * // seconds in a day
+				1.0 * // USD => GBP
+				100 // pences
+
+		ExpectJSON(out, []map[string]interface{}{
+			{
+				"org_guid":               guid,
+				"space_guid":             spaceGUID,
+				"price_in_pence_ex_vat":  expectedPrice,
+				"price_in_pence_inc_vat": expectedPrice * 1.20,
+			},
+		})
+	})
+
+	It("applies different currencies for different components", func() {
+		guid := "00000001-0000-0000-0000-000000000000"
+		spaceGUID := "00000001-0001-0000-0000-000000000000"
+
+		createScenario(
+			guid,
+			spaceGUID,
+			start,
+			[][]string{
+				{"GBP", "$time_in_seconds * 1.0"},
+				{"USD", "$time_in_seconds * 0.25"},
+			},
+		)
+
+		addCurrency("USD", 2, start)
+		addCurrency("USD", 4, start.Add(24*1*time.Hour))
+
+		var out interface{}
+
+		request(
+			"/organisations/"+guid+"/spaces",
+			start,
+			start.Add(3*24*time.Hour),
+			&out,
+		)
+
+		expectedPriceFirstComponent :=
+			24 * 3600 * // seconds in a day
+				1.0 * // price per second in component
+				(1 + 1 + 1) * // GBP => GBP each day
+				100 // pences
+
+		expectedPriceSecondComponent :=
+			24 * 3600 * // seconds in a day
+				0.25 * // price per second in component
+				(2 + 4 + 4) * // USD => GBP each day
+				100 // pences
+
+		expectedPrice := expectedPriceFirstComponent + expectedPriceSecondComponent
+
+		ExpectJSON(out, []map[string]interface{}{
+			{
+				"org_guid":               guid,
+				"space_guid":             spaceGUID,
+				"price_in_pence_ex_vat":  expectedPrice,
+				"price_in_pence_inc_vat": expectedPrice * 1.20,
+			},
+		})
+	})
+})
```

Comment:
I would like to see a few more test cases:
 - an event range is bigger than the request range
 - the event overlaps with a currency rate but that currency rate does not overlap with the request range
 - the event overlaps with a pricing plan but that pricing plan does not overlap with the currency rate

Reply:
> an event range is bigger than the request range

We are not testing event ranges. And the event range is always bigger, because we only insert a start event start

> the event overlaps with a currency rate but that currency rate does not overlap with the request range
> the event overlaps with a pricing plan but that pricing plan does not overlap with the currency rate

In both cases there must be always one currency and one plan overlapping, otherwise the report will report zero or empty.

We can add a test that says: does not report (or reports 0 usage) if the currency (or the plan) is missing.

Reply:
Yes, but you have to test these use cases I wrote, otherwise you don't validate all the date range intersections in the monetised_resources query. I suspect I can remove a couple of join conditions and these test would still pass.

Reply:
For the second and third use cases: sorry, I was not clear with these: in both cases there should be a currency_rate(2)/pricing plan(3) which overlaps, but there should be one currency_rate(2)/pricing_plan(3) which only overlaps with the request range. This is to test that the range intersections are correct in the query.